### PR TITLE
Export burp clients status to Prometheus DB

### PR DESCRIPTION
**Pull request by piligrim773**
Hello, we would like to export status of burp clients to Prometheus DB, what do you think about this approach

### CODE REVIEW

**Comment by dmitry-sinina** [comment]
@grke

**Comment by svalouch** [comment]
I'd really like to see Burp exposing its status to Prometheus. I don't speak for for @grke or anyone else working on this project, but adding the support for one specific monitoring system - however widely used it may be - to the codebase could open the path for others wanting their system supported as well, which puts additional burden on the developer.

Thus, I'd like to offer an alternative that I started (but have yet to finish) because of the above reasons: An exporter that connects to a running burp server, acting as a regular client. It works by using credentials generated using a regular burp client, but implements the protocol itself which I found easier than controlling a burp subprocess in monitoring mode (and a fun challenge). It also supports connecting to multiple burp servers simultaneously. The code can be found here: https://github.com/svalouch/burp_exporter

It is not finished / ready for production yet, and there are some challenges this method poses that are not present when spawning the exporter from the server directly as you are proposing:
* The burp server needs to allocate resources to the client, and the initial query could take a long time if the server is under a heavy load.
* The burp server does not remove clients that are no longer present in its `clientconfdir`, so I thought about adding something like an hourly reconnect or listening for a SIGHUP to refresh the connection.
* The client is added as a `restore_client` to the server config file (toplevel), which means that it could read other clients backups depending on the settings (this is more a "keep in mind" than a problem).

**Comment by grke** [comment]
Hello,

Thanks for both of your efforts!

While I haven't reviewed either of them properly yet, my immediate thought is that I don't really want to put third-party specific code into burp, and would prefer a generic mechanism.

**Comment by dmitry-sinina** [comment]
There are not so many stable well-known monitoring protocols, actually I know only two - prometheus and snmp.
I have seen https://github.com/svalouch/burp_exporter but we are trying to avoid additional processes  and dependencies on servers.

**Comment by grke** [comment]
Hello,

This patch fails to compile on my freebsd 12.0 test machine.
```
src/handy.c: In function 'set_tcp_defer_accept':
src/handy.c:260:32: error: 'TCP_DEFER_ACCEPT' undeclared (first use in this function); did you mean 'BIO_RR_ACCEPT'?
```

**Comment by grke** [comment]
Also, this happens with 'make runner':

```utest/test_conf.c: In function ‘check_default’:
utest/test_conf.c:10:2: warning: enumeration value ‘OPT_LISTEN_PROMETHEUS_EXPORTER’ not handled in switch [-Wswitch]
  switch(o)
  ^~~~~~
utest/test_conf.c:10:2: warning: enumeration value ‘OPT_PROMETHEUS_EXPORTER_REQUEST_TARGET’ not handled in switch [-Wswitch]
```

File: src/server/prometheus_exporter.c [hunk 1]
```diff
@@ -0,0 +1,499 @@
+﻿/**
```
Comment (grke) [inline src/server/prometheus_exporter.c:1]: None of this appears to have any test coverage.

**Comment by grke** [comment]
On consideration, I'm not a fan of the built-in approach, sorry. It's not something that I would be able to maintain and would probably break it frequently. It seems to me that a much better solution would be to read the burp monitor output, as this is the kind of thing that it was made for.

**Pull request closed without merging**